File: day10.py
with open('d.txt') as f:
    xs = f.readlines()

m = {')': '(', ']': '[', '}': '{', '>': '<'}
q = {')': 3, ']': 57, '}': 1197, '>': 25137}
qq = {'(': 1, '[': 2, '{': 3, '<': 4}

def unwind(t):
    s = 0
    while t:
        s *= 5
        s += qq[t.pop()]
    return s

def score(x):
    t = []
    for c in x:
        y = m.get(c)
        if y:
            if not t:
                return (q[c], None)
            elif t[-1] == y:
                t.pop()
            else:
                return (q[c], None)
        else:
            t.append(c)
    return (None, unwind(t))

s = 0
u = []
for x in xs:
    x = x.strip()
    g, h = score(x)
    if h is None:
        s += g
    else:  # g is None
        u.append(h)

print('part 1 =', s)
u.sort()
print('part 2 =', u[len(u)//2])

File: 2020-julia/day10.jl
lines = readlines("/home/xdavidliu/Documents/aoc/input10.txt")
nums = [parse(Int, x) for x in lines]
push!(nums, 0)
push!(nums, 3+maximum(nums))
sort!(nums)
n = length(nums)
cs = [0, 0, 0]
for i in 1:n-1
    d = nums[i+1]-nums[i]
    cs[d] += 1
end
println("part 1 = ", cs[1] * cs[3])  # 2201
dp = repeat([0], n+2)
dp[n] = 1
for i in reverse(1:n)
    k = i-1
    while 0 < k && nums[k] >= nums[i] - 3
        dp[k] += dp[i]
        k -= 1
    end
end
println("part 2 = ", dp[1])  # 169255295254528
